pulp_2to3_migration/app/plugin/deb/pulp_2to3_models.py

```diff
@@ -184,3 +190,307 @@ def create_pulp3_content(self):
+    component = django_models.TextField()
+    suite = django_models.TextField()
+
+    class Meta:
+        default_related_name = 'deb_component_detail_model'
+        # TODO: These are probably not the uniqueness constraints I want...
+        unique_together = ('component', 'distribution')
+
+    @classmethod
+    def pre_migrate_content_detail(cls, content_batch):
+        """
+        Pre-migrate Pulp 2 content with all the fields needed to create a Pulp 3 Content
+
+        Note that this pre_migrated_content_detail method will not just create
+        Pulp2DebComponent entries, but also Pulp2DebComponentPackage entries,
+        and Pulp2DebReleaseArchitecture entries, since they are all created from the same
+        Pulp 2 type (DebComponent).
+
+        For each Pulp2DebComponentPackage and each Pulp2DebComponent there will also be an
+        additional Pulp2Content created. Such Pulp2Content will have the same pulp2_id as
+        the original, but with a different pulp2_subid. That way we can record the one to
+        many Pulp 2 to Pulp 3 mapping later.
+
+        Args:
+             content_batch(list of Pulp2Content): pre-migrated generic data for Pulp 2 content.
+        """
+        pulp2_unit_map = {pulp2unit.pulp2_id: pulp2unit for pulp2unit in content_batch}
+        pulp2_ids = pulp2_unit_map.keys()
+        pulp2_unit_batch = pulp2_models.DebComponent.objects.filter(id__in=pulp2_ids)
+        component_units_to_save = []
+        component_package_units_to_save = []
+        release_architecture_units_to_save = []
+        pulp2_sub_records_to_save = []
+
+        for component_unit in pulp2_unit_batch:
+            pulp2_base_record = pulp2_unit_map[component_unit.id]
+            distribution = component_unit.distribution
+            codename = component_unit.release
+            component = component_unit.name
+            release = pulp2_models.DebRelease.objects.filter(
+                repoid=component_unit.repoid,
+                distribution=distribution,
+            ).first()
+            suite = release.suite
+            component_units_to_save.append(Pulp2DebComponent(
+                distribution=distribution,
+                codename=codename,
+                component=component,
+                suite=suite,
+                pulp2content=pulp2_base_record,
+            ))
+            architectures = set()
+            for package_id in component_unit.packages:
+                package_unit = pulp2_models.DebPackage.objects.filter(id__in=[package_id]).first()
+                package_relative_path = package_unit.filename
+                package_sha256 = package_unit.checksum
+                # TODO: Consider hashing the subid string to ensure it will fit in the pulp2_subid
+                pulp2_subid = (suite + codename + distribution + component
+                               + package_relative_path + package_sha256)
+
+                pulp2_sub_record = Pulp2Content(
```

Comment:
@goosemania I ended up not adding any `get_pulp2_subids` function and instead simply creating the "sub records" as part of the `pre_migrate_content_detail` function.

The reason for this, is that I only know how many "sub records" I will need (what there subids should be) once I run `pre_migrate_content_detail`. I am pretty confident it should not be a problem that they are created slightly later in the migration process than the other `Pulp2Content` units...

Reply:
I'm not sure how it works...
[Pulp2Content is already saved by this time](https://github.com/pulp/pulp-2to3-migration/blob/master/pulp_2to3_migration/app/pre_migration.py#L211-L217).
I haven't tested your PR, however I read it as you create new Pulp2Content which now have subids and which  you link to the detail model.
So what happens with the Pulp2Content created before that which would have subid = null?

Reply:
Every Pulp 2 `DebComponent` from Mongo is ultimately turned into exactly one Pulp 3 `ReleaseComponent`, as well as `n` `PackageReleaseComponent` and `m` `ReleaseArchitecture` entries.

The initial `Pulp2Content` entry (with `pulp2_subid = ''`) is associated with the (exactly one) Pulp 3 `ReleaseComponent`. I call this initial `Pulp2Content` entry the `pulp2_base_record` in the code. So far this is exactly the same as for every other type.

In addition, I am having `Pulp2DebComponent.pre_migrate_content_detail` create additional `Pulp2Content` units for each `Pulp2DebComponentPackage` and each `Pulp2DebReleaseArchitecture` I create. These additional `Pulp2Content` entries are given the same values as the corresponding `pulp2_base_record` for every field except the `pulp2_subid` field. The `pulp2_subid` field is generated from the `unique_together` fields of the relevant `Pulp2DebComponentPackage`/`Pulp2DebReleaseArchitecture`, which have a foreign key field on the newly created `Pulp2Content` `pulp2_sub_record`.

That is really all there is to it. Everything else appears to work as expected out of the box. There is now exactly one  `Pulp2Content` record for every Pulp3 entry that is eventually created and they are already correctly associated. The migration will mark them as migrated whenever the relevant Pulp3 content is created just like for everything else.

Reply:
![pulp_2to3_migration_flow](https://user-images.githubusercontent.com/2744127/107233168-9075be80-6a22-11eb-8a47-8ef04e779772.jpeg)
Not sure if this helps...

Reply:
Thanks for the detailed explanation!